drivers/wifi/esp8266/esp8266_offload.c

```diff
@@ -0,0 +1,737 @@
+		return 0;
+	}
+
+	ret = _sock_connect(dev, sock);
+
+	if (esp8266_socket_connected(sock) && sock->tx_pkt) {
+		k_work_submit_to_queue(&dev->workq, &sock->send_work);
+	}
+
+	esp8266_unlock(dev);
+
+	if (ret != -ETIMEDOUT && cb) {
+		cb(context, ret, user_data);
+	}
+
+	return ret;
+}
+
+static int esp8266_accept(struct net_context *context,
+			     net_tcp_accept_cb_t cb, s32_t timeout,
+			     void *user_data)
+{
+	return -ENOTSUP;
+}
+
+MODEM_CMD_DIRECT_DEFINE(on_cmd_tx_ready)
+{
+	struct esp8266_data *dev = CONTAINER_OF(data, struct esp8266_data,
+						cmd_handler_data);
+
+	k_sem_give(&dev->sem_tx_ready);
+	data->ret = len;
+}
+
+MODEM_CMD_DEFINE(on_cmd_send_ok)
+{
+	struct esp8266_data *dev = CONTAINER_OF(data, struct esp8266_data,
+						cmd_handler_data);
+
+	modem_cmd_handler_set_error(data, 0);
+	k_sem_give(&dev->sem_response);
+}
+
+MODEM_CMD_DEFINE(on_cmd_send_fail)
+{
+	struct esp8266_data *dev = CONTAINER_OF(data, struct esp8266_data,
+						cmd_handler_data);
+
+	modem_cmd_handler_set_error(data, -EIO);
+	k_sem_give(&dev->sem_response);
+}
+
+static int _sock_send(struct esp8266_data *dev, struct esp8266_socket *sock)
+{
+	char cmd_buf[64], addr_str[NET_IPV4_ADDR_LEN];
+	int ret, write_len, pkt_len;
+	struct modem_cmd cmds[] = {
+		MODEM_CMD_DIRECT(">", on_cmd_tx_ready),
+		MODEM_CMD("SEND OK", on_cmd_send_ok, 0U, ""),
+		MODEM_CMD("SEND FAIL", on_cmd_send_fail, 0U, ""),
+	};
+
+	if (!esp8266_flag_is_set(dev, EDF_STA_CONNECTED)) {
+		return -ENETUNREACH;
+	}
+
+	pkt_len = net_pkt_get_len(sock->tx_pkt);
+
+	LOG_DBG("link %d, len %d", sock->link_id, pkt_len);
+
+	if (sock->ip_proto == IPPROTO_TCP) {
+		snprintk(cmd_buf, sizeof(cmd_buf),
+			 "AT+CIPSEND=%d,%d", sock->link_id, pkt_len);
+	} else {
+		net_addr_ntop(sock->dst.sa_family,
+			      &net_sin(&sock->dst)->sin_addr,
+			      addr_str, sizeof(addr_str));
+		snprintk(cmd_buf, sizeof(cmd_buf),
+			 "AT+CIPSEND=%d,%d,\"%s\",%d",
+			 sock->link_id, pkt_len, addr_str,
+			 ntohs(net_sin(&sock->dst)->sin_port));
+	}
+
+	k_sem_take(&dev->cmd_handler_data.sem_tx_lock, K_FOREVER);
+	k_sem_reset(&dev->sem_tx_ready);
+
+	ret = modem_cmd_send_nolock(&dev->mctx.iface, &dev->mctx.cmd_handler,
+			     NULL, 0, cmd_buf, &dev->sem_response,
+			     ESP_CMD_TIMEOUT);
+	if (ret < 0) {
+		LOG_DBG("Failed to send command");
+		goto out;
+	}
+
+	ret = modem_cmd_handler_update_cmds(&dev->cmd_handler_data,
+					    cmds, ARRAY_SIZE(cmds),
+					    true);
+	if (ret < 0) {
+		goto out;
+	}
+
+	/*
+	 * After modem handlers have been updated the receive buffer
+	 * needs to be processed again since there might now be a match.
+	 */
+	k_sem_give(&dev->iface_data.rx_sem);
+
+	/* Wait for '>' */
+	ret = k_sem_take(&dev->sem_tx_ready, 5000);
+	if (ret < 0) {
+		LOG_DBG("Timeout waiting for tx");
+		goto out;
+	}
+
+	while (pkt_len) {
+		write_len = MIN(pkt_len, sizeof(dev->pkt_buf));
+		net_pkt_read(sock->tx_pkt, dev->pkt_buf, write_len);
+		dev->mctx.iface.write(&dev->mctx.iface, dev->pkt_buf,
```

Comment:
To help with future debugging could you add something like:

```
		#if defined(CONFIG_MODEM_CONTEXT_VERBOSE_DEBUG)
				LOG_HEXDUMP_DBG(dev->pkt_buf, write_len, "SENT:");
		#endif
```

Reply:
Fixed in https://github.com/zephyrproject-rtos/zephyr/pull/23007